File: app/src/main/java/ir/arapp/arappmain/View/Fragments/ForgetPassValidateFragment.java
package ir.arapp.arappmain.View.Fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import java.util.Objects;
import ir.arapp.arappmain.R;
import ir.arapp.arappmain.Util.Services.FragmentManager;
import ir.arapp.arappmain.Util.Services.HideShowKeyboard;
import ir.arapp.arappmain.Util.Services.SnackBarMessage;
import ir.arapp.arappmain.Util.Services.SnackBarToast;
import ir.arapp.arappmain.databinding.FragmentValidateForgetPassBinding;
import ir.arapp.arappmain.viewmodel.ForgetPassViewModel;

public class ForgetPassValidateFragment extends Fragment implements SnackBarMessage, FragmentManager
{

//    region Variables
    FragmentValidateForgetPassBinding fragmentValidateForgetPassBinding;
    private SnackBarToast snackBarToast;
    private HideShowKeyboard hideShowKeyboard;
//    endregion

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
//        Inflate the layout for this fragment
        fragmentValidateForgetPassBinding = FragmentValidateForgetPassBinding.inflate(inflater, container, false);
//        Set view model
        ForgetPassViewModel forgetPassValidateViewModel = ViewModelProviders.of(requireActivity()).get(ForgetPassViewModel.class);
        fragmentValidateForgetPassBinding.setViewModel(forgetPassValidateViewModel);
        forgetPassValidateViewModel.snackBarMessage = this;
        forgetPassValidateViewModel.fragmentManager = this;
//        Life cycle owner
        fragmentValidateForgetPassBinding.setLifecycleOwner(this);
//        Hooks
        snackBarToast = new SnackBarToast(fragmentValidateForgetPassBinding.getRoot());
        hideShowKeyboard = new HideShowKeyboard(getContext(), fragmentValidateForgetPassBinding.getRoot());
//        Toolbar
        ((AppCompatActivity) requireActivity()).setSupportActionBar(fragmentValidateForgetPassBinding.forgetPassToolbar);
        Objects.requireNonNull(((AppCompatActivity) requireActivity()).getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        fragmentValidateForgetPassBinding.forgetPassToolbar.setNavigationOnClickListener(view1 -> onNavigateUp());
//        Pin view animation
        fragmentValidateForgetPassBinding.pinView.setAnimationEnable(true);
        pinViewTryAgainLineColor();
//        Return view
        return fragmentValidateForgetPassBinding.getRoot();
    }

//    region methods
//    Back button navigation
    private void onNavigateUp()
    {
        requireActivity().onBackPressed();
    }
//    Set methods for specific function
    @Override
    public void setFunction(String type)
    {
        switch (type)
        {
            case "error":
                fragmentValidateForgetPassBinding.pinView.setLineColor(
                        ResourcesCompat.getColor(getResources(), R.color.notificationColorRed, requireActivity().getTheme()));
                break;
            case "resend":
                fragmentValidateForgetPassBinding.timer.setTextColor(getResources().getColor(R.color.colorAccentDark));
                fragmentValidateForgetPassBinding.pinView.setText("");
                fragmentValidateForgetPassBinding.pinView.setLineColor(
                        ResourcesCompat.getColor(getResources(), R.color.colorPrimaryDark, requireActivity().getTheme()));
                break;
        }
    }
//    Pin View change Line Color
    private void pinViewTryAgainLineColor()
    {
        fragmentValidateForgetPassBinding.pinView.addTextChangedListener(new TextWatcher()
        {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2)
            {
            }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2)
            {
                fragmentValidateForgetPassBinding.pinView.setLineColor(
                        ResourcesCompat.getColor(getResources(), R.color.colorPrimaryDark, requireActivity().getTheme()));
            }
            @Override
            public void afterTextChanged(Editable editable)
            {
            }
        });
    }
//    Fragment navigation
    @Override
    public void navigateToFragment(String message)
    {
//        Nav Controller
        final NavController navController = Navigation.findNavController(fragmentValidateForgetPassBinding.getRoot());

        switch (message)
        {
            case "forgetPass":
                navController.navigate(R.id.action_forgetPassValidateFragment_to_forgetPassFragment);
                hideShowKeyboard.hideKeyboardFrom(true);
                break;
            case "phone":
                navController.navigate(R.id.action_forgetPassValidateFragment_to_forgetPassPhoneFragment);
                requireActivity().getViewModelStore().clear();
                hideShowKeyboard.hideKeyboardFrom(true);
                break;
        }
    }
//    Error/success message
    @Override
    public void onSuccess(String message)
    {
        snackBarToast.snackBarShortTime(message);
    }
    @Override
    public void onFailure(String message)
    {
        snackBarToast.snackBarShortTime(message);
    }
//    endregion
}

File: app/src/main/java/ir/arapp/arappmain/view/fragments/ForgetPassValidateFragment.kt
package ir.arapp.arappmain.view.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import ir.arapp.arappmain.R
import ir.arapp.arappmain.databinding.FragmentValidateForgetPassBinding
import ir.arapp.arappmain.util.services.FragmentManager
import ir.arapp.arappmain.util.services.HideShowKeyboard
import ir.arapp.arappmain.util.services.SnackBarMessage
import ir.arapp.arappmain.util.services.SnackBarToast
import ir.arapp.arappmain.viewmodel.ForgetPassViewModel
import java.util.*

class ForgetPassValidateFragment : Fragment(), SnackBarMessage, FragmentManager {
    //    region Variables
    var _fragmentValidateForgetPassBinding: FragmentValidateForgetPassBinding? = null
    val fragmentValidateForgetPassBinding get() = _fragmentValidateForgetPassBinding!!
    private var snackBarToast: SnackBarToast? = null
    private var hideShowKeyboard: HideShowKeyboard? = null

    override fun onDestroyView() {
        super.onDestroyView()
        _fragmentValidateForgetPassBinding = null
    }
    //    endregion
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
//        Inflate the layout for this fragment
        _fragmentValidateForgetPassBinding = DataBindingUtil.inflate(
            inflater,
            R.layout.fragment_validate_forget_pass,
            container,
            false
        )
        //        Set view model
        val forgetPassValidateViewModel = ViewModelProvider(requireActivity()).get(
            ForgetPassViewModel::class.java
        )
        fragmentValidateForgetPassBinding.setViewModel(forgetPassValidateViewModel)
        forgetPassValidateViewModel.snackBarMessage = this
        forgetPassValidateViewModel.fragmentManager = this
        //        Life cycle owner
        fragmentValidateForgetPassBinding.setLifecycleOwner(this)
        //        Hooks
        snackBarToast = SnackBarToast(fragmentValidateForgetPassBinding.getRoot())
        hideShowKeyboard = HideShowKeyboard(requireContext(), fragmentValidateForgetPassBinding.getRoot())
        //        Toolbar
        (requireActivity() as AppCompatActivity).setSupportActionBar(
            fragmentValidateForgetPassBinding.forgetPassToolbar
        )
        Objects.requireNonNull((requireActivity() as AppCompatActivity).supportActionBar)!!
            .setDisplayHomeAsUpEnabled(true)
        fragmentValidateForgetPassBinding.forgetPassToolbar.setNavigationOnClickListener { view1: View? -> onNavigateUp() }
        //        Pin view animation
        fragmentValidateForgetPassBinding.pinView.setAnimationEnable(true)
        pinViewTryAgainLineColor()
        //        Return view
        return fragmentValidateForgetPassBinding.getRoot()
    }

    //    region methods
    //    Back button navigation
    private fun onNavigateUp() {
        requireActivity().onBackPressed()
    }

    //    Set methods for specific function
    override fun setFunction(type: String?) {
        when (type) {
            "error" -> fragmentValidateForgetPassBinding!!.pinView.setLineColor(
                ResourcesCompat.getColor(
                    resources,
                    R.color.notificationColorRed,
                    requireActivity().theme
                )
            )
            "editPhone" -> {
                fragmentValidateForgetPassBinding!!.editPhone.isEnabled = true
                fragmentValidateForgetPassBinding!!.editPhone.setTextColor(resources.getColor(R.color.colorAccent))
            }
            "resend" -> {
                fragmentValidateForgetPassBinding!!.pinView.setText("")
                fragmentValidateForgetPassBinding!!.editPhone.isEnabled = false
                fragmentValidateForgetPassBinding!!.editPhone.setTextColor(resources.getColor(R.color.disable))
                fragmentValidateForgetPassBinding!!.pinView.setLineColor(
                    ResourcesCompat.getColor(
                        resources,
                        R.color.colorPrimaryDark,
                        requireActivity().theme
                    )
                )
            }
        }
    }

    //    Pin View change Line Color
    private fun pinViewTryAgainLineColor() {
        fragmentValidateForgetPassBinding!!.pinView.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                fragmentValidateForgetPassBinding!!.pinView.setLineColor(
                    ResourcesCompat.getColor(
                        resources,
                        R.color.colorPrimaryDark,
                        requireActivity().theme
                    )
                )
            }

            override fun afterTextChanged(editable: Editable) {}
        })
    }

    //    Fragment navigation
    override fun navigateToFragment(message: String?) {
//        Nav Controller
        val navController = Navigation.findNavController(
            fragmentValidateForgetPassBinding!!.root
        )
        when (message) {
            "forgetPass" -> {
                navController.navigate(R.id.action_forgetPassValidateFragment_to_forgetPassFragment)
                hideShowKeyboard!!.hideKeyboardFrom(true)
            }
            "phone" -> {
                navController.navigate(R.id.action_forgetPassValidateFragment_to_forgetPassPhoneFragment)
                hideShowKeyboard!!.hideKeyboardFrom(true)
            }
        }
    }

    //    Error/success message
    override fun onSuccess(message: String?) {
        snackBarToast!!.snackBarShortTime(message)
    }

    override fun onFailure(message: String?) {
        snackBarToast!!.snackBarShortTime(message)
    } //    endregion
}